File: crates/deepsec-cli/src/commands/report.rs
use crate::config_loader::Context;
use anyhow::Result;
use clap::Args as ClapArgs;
use colored::Colorize;
use deepsec_core::{
    FileRecord, Finding, RevalidationVerdict, Severity, report_csv_path, report_json_path,
    report_md_path, store::load_all_file_records,
};

#[derive(Debug, ClapArgs)]
pub struct Args {
    #[arg(long)]
    pub project_id: String,
    /// Minimum severity to include.
    #[arg(long)]
    pub min_severity: Option<String>,
    /// Only include findings produced by this run.
    #[arg(long)]
    pub run_id: Option<String>,
    /// Exclude findings revalidated as false-positive / fixed.
    #[arg(long)]
    pub real_only: bool,
}

pub fn run(args: Args, ctx: &Context) -> Result<()> {
    let records = load_all_file_records(&ctx.data_root, &args.project_id)?;
    let min_sev = args
        .min_severity
        .as_deref()
        .map(parse_severity)
        .transpose()?
        .unwrap_or(Severity::Low);

    let mut rows: Vec<(String, &FileRecord, &Finding)> = Vec::new();
    for rec in &records {
        for f in &rec.findings {
            if f.severity.rank() < min_sev.rank() {
                continue;
            }
            if let Some(rid) = &args.run_id {
                if f.produced_by_run_id.as_deref() != Some(rid.as_str()) {
                    continue;
                }
            }
            if args.real_only {
                if let Some(rv) = &f.revalidation {
                    if matches!(
                        rv.verdict,
                        RevalidationVerdict::FalsePositive | RevalidationVerdict::Fixed
                    ) {
                        continue;
                    }
                }
            }
            rows.push((rec.file_path.clone(), rec, f));
        }
    }
    rows.sort_by(|a, b| b.2.severity.rank().cmp(&a.2.severity.rank()));

    let md_path = report_md_path(&ctx.data_root, &args.project_id, args.run_id.as_deref())?;
    let json_path = report_json_path(&ctx.data_root, &args.project_id, args.run_id.as_deref())?;
    let csv_path = report_csv_path(&ctx.data_root, &args.project_id, args.run_id.as_deref())?;
    if let Some(p) = md_path.parent() {
        fs_err::create_dir_all(p)?;
    }
    fs_err::write(&md_path, render_markdown(&args.project_id, &rows))?;
    fs_err::write(&json_path, render_json(&rows)?)?;
    fs_err::write(&csv_path, render_csv(&rows)?)?;

    println!(
        "{} findings={} → {}, {}, {}",
        "report".bold().green(),
        rows.len(),
        md_path.display(),
        json_path.display(),
        csv_path.display()
    );
    Ok(())
}

fn parse_severity(s: &str) -> Result<Severity> {
    Ok(match s.to_uppercase().as_str() {
        "CRITICAL" => Severity::Critical,
        "HIGH" => Severity::High,
        "HIGH_BUG" => Severity::HighBug,
        "MEDIUM" => Severity::Medium,
        "BUG" => Severity::Bug,
        "LOW" => Severity::Low,
        _ => anyhow::bail!("unknown severity: {s}"),
    })
}

fn render_markdown(project_id: &str, rows: &[(String, &FileRecord, &Finding)]) -> String {
    use std::fmt::Write;
    let mut out = String::new();
    writeln!(&mut out, "# deepsec report — {project_id}\n").ok();
    writeln!(&mut out, "Total findings: {}\n", rows.len()).ok();
    for (path, _rec, f) in rows {
        writeln!(
            &mut out,
            "## [{}] {} — `{}`",
            f.severity.as_str(),
            f.title,
            f.vuln_slug
        )
        .ok();
        writeln!(&mut out, "- file: `{}`", path).ok();
        writeln!(&mut out, "- lines: {:?}", f.line_numbers).ok();
        writeln!(&mut out, "- confidence: {:?}", f.confidence).ok();
        if let Some(r) = &f.revalidation {
            writeln!(&mut out, "- revalidation: {:?}", r.verdict).ok();
        }
        if let Some(t) = &f.triage {
            writeln!(
                &mut out,
                "- triage: priority={:?} exploitability={:?} impact={:?}",
                t.priority, t.exploitability, t.impact
            )
            .ok();
        }
        writeln!(&mut out, "\n{}\n", f.description).ok();
        writeln!(&mut out, "**Fix:** {}\n", f.recommendation).ok();
    }
    out
}

fn render_json(rows: &[(String, &FileRecord, &Finding)]) -> Result<String> {
    let payload: Vec<_> = rows
        .iter()
        .map(|(path, _, f)| {
            serde_json::json!({
                "filePath": path,
                "finding": f,
            })
        })
        .collect();
    Ok(serde_json::to_string_pretty(&payload)?)
}

fn render_csv(rows: &[(String, &FileRecord, &Finding)]) -> Result<String> {
    let mut wtr = csv::Writer::from_writer(Vec::new());
    wtr.write_record(["file", "severity", "slug", "lines", "title", "confidence", "verdict"])?;
    for (path, _, f) in rows {
        let lines: Vec<String> = f.line_numbers.iter().map(|n| n.to_string()).collect();
        wtr.write_record([
            path,
            f.severity.as_str(),
            &f.vuln_slug,
            &lines.join(";"),
            &f.title,
            &format!("{:?}", f.confidence),
            &f.revalidation
                .as_ref()
                .map(|r| format!("{:?}", r.verdict))
                .unwrap_or_default(),
        ])?;
    }
    Ok(String::from_utf8(wtr.into_inner()?)?)
}

File: internal/cli/commands/report.go
package commands

import (
	"encoding/csv"
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"sort"
	"strconv"
	"strings"

	"github.com/noeljackson/deepsec/internal/cli"
	"github.com/noeljackson/deepsec/internal/core"
	"github.com/spf13/cobra"
)

// NewReportCmd writes markdown + JSON + CSV reports, or a single
// compliance-formatted JSON for `--format soc2|ssdf`.
func NewReportCmd(loader func() (*cli.Context, error)) *cobra.Command {
	var projectID, minSeverity, runID, format, output, verifyPath string
	var realOnly bool
	cmd := &cobra.Command{
		Use:   "report",
		Short: "Render a project report (markdown + JSON + CSV) or compliance-formatted output",
		RunE: func(cmd *cobra.Command, _ []string) error {
			if verifyPath != "" {
				return runComplianceVerify(verifyPath)
			}
			ctx, err := loader()
			if err != nil {
				return err
			}
			records, err := ctx.DataRoot.LoadAllFileRecords(projectID)
			if err != nil {
				return err
			}

			if format == "soc2" || format == "ssdf" {
				return runComplianceReport(ctx, projectID, records, format, output)
			}
			minSev := core.SeverityLow
			if minSeverity != "" {
				minSev = core.Severity(strings.ToUpper(minSeverity))
				if minSev.Rank() == 0 {
					return fmt.Errorf("unknown severity: %s", minSeverity)
				}
			}
			rows := collectRows(records, minSev, runID, realOnly)
			sort.SliceStable(rows, func(i, j int) bool {
				return rows[i].finding.Severity.Rank() > rows[j].finding.Severity.Rank()
			})

			mdPath, err := ctx.DataRoot.ReportMDPath(projectID, runID)
			if err != nil {
				return err
			}
			jsonPath, err := ctx.DataRoot.ReportJSONPath(projectID, runID)
			if err != nil {
				return err
			}
			csvPath, err := ctx.DataRoot.ReportCSVPath(projectID, runID)
			if err != nil {
				return err
			}
			if err := os.MkdirAll(filepath.Dir(mdPath), 0o755); err != nil {
				return err
			}
			if err := os.WriteFile(mdPath, []byte(renderMarkdown(projectID, rows)), 0o644); err != nil {
				return err
			}
			body, err := renderJSON(rows)
			if err != nil {
				return err
			}
			if err := os.WriteFile(jsonPath, body, 0o644); err != nil {
				return err
			}
			if err := writeCSV(csvPath, rows); err != nil {
				return err
			}
			fmt.Printf("report findings=%d → %s, %s, %s\n",
				len(rows), mdPath, jsonPath, csvPath)
			return nil
		},
	}
	cmd.Flags().StringVar(&projectID, "project-id", "", "Project id (required unless --verify is used)")
	cmd.Flags().StringVar(&minSeverity, "min-severity", "", "Minimum severity to include (default 3 outputs only)")
	cmd.Flags().StringVar(&runID, "run-id", "", "Only include findings produced by this run")
	cmd.Flags().BoolVar(&realOnly, "real-only", false, "Drop FP/Fixed findings (default 3 outputs only)")
	cmd.Flags().StringVar(&format, "format", "", "Output format: empty for the default markdown+JSON+CSV trio, or 'soc2'/'ssdf' for compliance-formatted output")
	cmd.Flags().StringVar(&output, "output", "", "Output file path for compliance-formatted output (default: data/<project>/reports/compliance.json)")
	cmd.Flags().StringVar(&verifyPath, "verify", "", "Verify a previously generated compliance report against DEEPSEC_REPORT_SIGNING_KEY")
	return cmd
}

func runComplianceReport(ctx *cli.Context, projectID string, records []*core.FileRecord, format, output string) error {
	if projectID == "" {
		return fmt.Errorf("--project-id is required for compliance reports")
	}
	runs, err := ctx.DataRoot.ListRuns(projectID)
	if err != nil {
		return err
	}
	flat := make([]core.RunMeta, 0, len(runs))
	for _, r := range runs {
		if r != nil {
			flat = append(flat, *r)
		}
	}
	rep, err := BuildComplianceReport(format, projectID, records, flat)
	if err != nil {
		return err
	}
	if output == "" {
		root, err := ctx.DataRoot.ReportJSONPath(projectID, "")
		if err != nil {
			return err
		}
		output = filepath.Join(filepath.Dir(root), "compliance."+format+".json")
	}
	if err := writeComplianceReport(rep, output); err != nil {
		return err
	}
	signed := ""
	if rep.Signature != nil {
		signed = fmt.Sprintf(" (signed: %s key_hint=%s)", rep.Signature.Algorithm, rep.Signature.KeyHint)
	}
	fmt.Printf("compliance report (%s) findings=%d → %s%s\n", format, len(rep.Findings), output, signed)
	return nil
}

func runComplianceVerify(path string) error {
	key := os.Getenv("DEEPSEC_REPORT_SIGNING_KEY")
	if key == "" {
		return fmt.Errorf("DEEPSEC_REPORT_SIGNING_KEY required for --verify")
	}
	if err := verifyComplianceFile(path, []byte(key)); err != nil {
		return err
	}
	fmt.Printf("compliance report verified: %s\n", path)
	return nil
}

type row struct {
	path    string
	record  *core.FileRecord
	finding *core.Finding
}

func collectRows(records []*core.FileRecord, minSev core.Severity, runID string, realOnly bool) []row {
	var rows []row
	for _, rec := range records {
		for i := range rec.Findings {
			f := &rec.Findings[i]
			if f.Severity.Rank() < minSev.Rank() {
				continue
			}
			if runID != "" && f.ProducedByRunID != runID {
				continue
			}
			if realOnly && f.Revalidation != nil {
				v := f.Revalidation.Verdict
				if v == core.VerdictFalsePositive || v == core.VerdictFixed {
					continue
				}
			}
			rows = append(rows, row{path: rec.FilePath, record: rec, finding: f})
		}
	}
	return rows
}

func renderMarkdown(projectID string, rows []row) string {
	var b strings.Builder
	fmt.Fprintf(&b, "# deepsec report — %s\n\n", projectID)
	fmt.Fprintf(&b, "Total findings: %d\n\n", len(rows))
	for _, r := range rows {
		f := r.finding
		fmt.Fprintf(&b, "## [%s] %s — `%s`\n", f.Severity, f.Title, f.VulnSlug)
		fmt.Fprintf(&b, "- file: `%s`\n", r.path)
		fmt.Fprintf(&b, "- lines: %v\n", f.LineNumbers)
		fmt.Fprintf(&b, "- confidence: %s\n", f.Confidence)
		if f.Revalidation != nil {
			fmt.Fprintf(&b, "- revalidation: %s\n", f.Revalidation.Verdict)
		}
		if f.Triage != nil {
			fmt.Fprintf(&b, "- triage: priority=%s exploitability=%s impact=%s\n",
				f.Triage.Priority, f.Triage.Exploitability, f.Triage.Impact)
		}
		fmt.Fprintf(&b, "\n%s\n\n", f.Description)
		fmt.Fprintf(&b, "**Fix:** %s\n\n", f.Recommendation)
	}
	return b.String()
}

func renderJSON(rows []row) ([]byte, error) {
	type payload struct {
		FilePath string        `json:"filePath"`
		Finding  *core.Finding `json:"finding"`
	}
	out := make([]payload, 0, len(rows))
	for _, r := range rows {
		out = append(out, payload{FilePath: r.path, Finding: r.finding})
	}
	return json.MarshalIndent(out, "", "  ")
}

func writeCSV(path string, rows []row) error {
	f, err := os.Create(path)
	if err != nil {
		return err
	}
	defer f.Close()
	w := csv.NewWriter(f)
	defer w.Flush()
	_ = w.Write([]string{"file", "severity", "slug", "lines", "title", "confidence", "verdict"})
	for _, r := range rows {
		lines := make([]string, 0, len(r.finding.LineNumbers))
		for _, n := range r.finding.LineNumbers {
			lines = append(lines, strconv.Itoa(n))
		}
		verdict := ""
		if r.finding.Revalidation != nil {
			verdict = string(r.finding.Revalidation.Verdict)
		}
		_ = w.Write([]string{
			r.path,
			string(r.finding.Severity),
			r.finding.VulnSlug,
			strings.Join(lines, ";"),
			r.finding.Title,
			string(r.finding.Confidence),
			verdict,
		})
	}
	return nil
}
